cpp/source/deal.II/electrochemical_physics.templates.h

```diff
@@ -0,0 +1,257 @@
+                   fe_values[solid_potential].value(j, q) -
+               fe_values[solid_potential].value(i, q) *
+                   fe_values[liquid_potential].value(j, q) -
+               fe_values[liquid_potential].value(i, q) *
+                   fe_values[solid_potential].value(j, q) +
+               fe_values[liquid_potential].value(i, q) *
+                   fe_values[liquid_potential].value(j, q)) *
+              fe_values.JxW(q);
+          cell_mass_matrix(i, j) += mass_matrix_val;
+          cell_system_matrix(i, j) +=
+              mass_matrix_val +
+              time_step *
+                  // Stiffness matrix terms
+                  (solid_phase_diffusion_coefficient_values[q] *
+                       (fe_values[solid_potential].gradient(i, q) *
+                        fe_values[solid_potential].gradient(j, q)) +
+                   liquid_phase_diffusion_coefficient_values[q] *
+                       (fe_values[liquid_potential].gradient(i, q) *
+                        fe_values[liquid_potential].gradient(j, q)) +
+                   faradaic_reaction_coefficient_values[q] *
+                       ((fe_values[solid_potential].value(i, q) *
+                         fe_values[solid_potential].value(j, q)) -
+                        (fe_values[liquid_potential].value(i, q) *
+                         fe_values[solid_potential].value(j, q)) -
+                        (fe_values[solid_potential].value(i, q) *
+                         fe_values[liquid_potential].value(j, q)) +
+                        (fe_values[liquid_potential].value(i, q) *
+                         fe_values[liquid_potential].value(j, q)))) *
+                  fe_values.JxW(q);
+        }
+      }
+
+    // Fill in the global matrices.
+    cell->get_dof_indices(local_dof_indices);
+    this->constraint_matrix.distribute_local_to_global(
+        cell_system_matrix, cell_rhs, local_dof_indices, this->system_matrix,
+        this->system_rhs, inhomogeneous_bc);
+    // TODO is this correct? Do we have constraints on the mass matrix?
+    for (unsigned int i = 0; i < dofs_per_cell; ++i)
+      for (unsigned int j = 0; j < dofs_per_cell; ++j)
+        this->mass_matrix.add(local_dof_indices[i], local_dof_indices[j],
+                              cell_mass_matrix(i, j));
+  }
+
+  // Add the null space. This wouldn't be necessary if we were using a
+  // hp::DoFHandler object instead of a DoFHandler object but
+  // hp::DoFHandler does not work with MPI.
+  dealii::types::global_dof_index const n_dofs = this->dof_handler->n_dofs();
+  for (unsigned int i = 0; i < n_dofs; ++i)
+    if (std::fabs(this->system_matrix.diag_element(i)) < 1e-100)
+      this->system_matrix.diag_element(i) = 1.;
+
+  // Apply Neumann boundary condition on the cathode (constant current
+  // charge).
+  if (electrochemical_parameters->charge_type == ConstantCurrent)
+  {
+    double const time_step = electrochemical_parameters->time_step;
```

Comment:
I don't see how that's a problem. This class builds the system matrix and the system matrix needs to know the time step. Remember, I told you that this can be solved by writing a mass matrix ```linear_operator```, a stiffness matrix ```linear_operator```, and moving the constraint matrix to be applied just before the solve but you said I can do it the way I wanted :p